pkg/client/apiutil/apimachinery.go

```diff
@@ -169,3 +144,23 @@ func createRestConfig(gvk schema.GroupVersionKind, isUnstructured bool, baseConf
 
 	return cfg
 }
+
+// protoMessage is implemented by protobuf messages (of all libraries)
+type protoMessage interface {
+	ProtoMessage()
+}
+
+var protoMessageType = reflect.TypeOf(new(protoMessage)).Elem()
+
+// haveProtoDefinition returns true if the go type for the specified gvk support protobuf encoding.
+func haveProtoDefinition(scheme *runtime.Scheme, gvk schema.GroupVersionKind) bool {
+	gvkType, found := scheme.AllKnownTypes()[gvk]
+	if !found {
+		return false
+	}
+	// We normally get the raw struct e.g. v1.Pod, not &v1.Pod
+	if gvkType.Kind() == reflect.Struct {
+		gvkType = reflect.PtrTo(gvkType)
+	}
+	return gvkType.Implements(protoMessageType)
```

Comment:
I'm afraid the gvk type has proto definition in scheme doesn't mean the API Server could accept its requests via Protobuf. Actually for a gvk type defined by CRD, API Server can only accept JSON requests even if we implement its proto types in project.

Reply:
Do you know if there's a way to know if the apiserver supports proto (for a given kind)?  I think we need "destination-type-is-proto and apiserver-supports-proto", because if someone queries a PodList into an UnstructuredList we still want to not use proto (?).

BTW the reason that I'm looking at this code is that I'd like to eliminate most of the imports of the kubernetes API types, for controllers that generally just use unstructured.Unstructured.

Reply:
You can see this PR [Use application/vnd.kubernetes.protobuf as content-type if possible](https://github.com/kubernetes-sigs/controller-runtime/pull/1149) I merged before, which brings Protobuf support into controller-runtime.

```
// Currently only enabled for built-in resources which are guaranteed to implement Protocol Buffers.
// For custom resources, CRDs can not support Protocol Buffers but Aggregated API can.
// See doc: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility
```

So if you want to extend the no-built-in types to use Protobuf, you have to find out whether they are defined by CRD or Aggregated API, which is hard to check. Since all types created via kubebuilder/operator-sdk are in CRD definition, I think the built-in types are enough for most ppl using controller-runtime.

WDYT @alvaroaleman

Reply:
Good point, I guess it is possible (and something I would love to try out!) to use proto to define CRDs locally.

I tweaked the logic therefore, there's a list of built-in types, and we have a unit test to make sure that the core kubernetes types are recognized appropriately.  This should detect if there's a new built-in library added upstream, and I added a comment so that it's obvious how to add the newly-added built-in group.

This also means that we can continue to support AddToProtobufScheme.